api/nakadi-event-bus-api.yaml

```diff
@@ -0,0 +1,1232 @@
+      reception, but they in general MIGHT be set by the client.
+    properties:
+      metadata:
+        type: object
+        properties:
+          eid:
+            description: |
+              Identifier of this Event.
+
+              Clients are allowed to generate this and this SHOULD be guaranteed to be unique from the
+              perspective of the producer. Consumers MIGHT use this value to assert uniqueness of reception
+              of the Event.
+
+              If not present upon reception, Nakadi will generate and fill this with a unique value.
+              (TBD: enforce UUID formatif set by the client?)
+            type: string
+            example: '105a76d8-db49-4144-ace7-e683e8f4ba46'
+          event_type:
+            description: |
+              The EventType of this Event. This is enriched by Nakadi on reception of the Event based on the
+              endpoint where the Producer sent the Event to.
+
+              If provided MUST match the endpoint. Failure to do so will cause rejection of the Event.
+            type: string
+            example: 'pennybags:payment-business-event'
+          occurred_at:
+            description: |
+              Timestamp of creation of the Event generated by the producer. Presence is optional but strongly
+              recommended for producers.
+            type: string
+            format: date-time
+          received_at:
+            type: string
+            description: |
+              Timestamp of the reception of the Event by Nakadi. This is enriched upon reception of the
+              Event.
+              If set by the producer Event will be rejected.
+            format: date-time
+          parent_eid:
+            type: string
+            format: uuid
+            description: |
+              Event identifier of the Event that caused the generation of this Event.
+              Set by the procuder.
+            example: '105a76d8-db49-4144-ace7-e683e8f4ba46'
+          root_eid:
+            type: string
+            format: uuid
+            description: |
+              Event identifier of the root of the tree of Events (from the relation constructed from
+              parent-eid)
+            example: '105a76d8-db49-4144-ace7-e683e8f4ba46'
+          flow_id:
+            description: |
+              The flow-id of the producer of this Event. As this is usually a HTTP header, this is enriched
+              from the header into the metadata by Nakadi to avoid clients having to explicitly copy this.
+            type: string
+            example: 'JAh6xH4OQhCJ9PutIV_RYw'
+
+        additionalProperties: false
+
+    required:
+      - metadata
+
+  BusinessEvent:
+    type: object
+    description: |
+      A Business Event.
+
+      Usually represents a status transition in a Business process.
+    allOf:
+      - $ref: '#/definitions/EventMetadata'
+    additionalProperties: true
+
+  DataChangeEventQualifier:
```

Comment:
@aikej it is a very important and difficult discussion. I would compleatly support your point of view about the resources exposed over the RESTful APIs. The internal structures should be of no concern for anybody outside the team.

The problem that is quite difficult to solve though (and it is of cause out of scope for Nakadi actually), is a more generic problem: how would a service ensure, that what is sent over the bus, is actually what was persisted. If the queue is not the only (and the primary) persistence layer, then any `queue` <-> `persistence layer` operation should be performed in a 2PC (two phase commit), that would be very problematic, even if Nakadi would implement _prepared transactions_. Not every persistence solution implements transactions to support the 2PC. 

From the other hand, if we would use some kind of a _driver level_ persistence layer log extractor, that would eliminate the need to write to two systems in parallel, but would extract the results of the persistence operations in the form of the decoded log, that is automatically delivered into the Nakadi system. One could give much higher consistency guarantees for the whole system.

I would be really glad to discuss this important topic in a more structured way, then using this API review.